### ISSUE TITLE
LIST .... FROM ... IN ... not working

### ISSUE THREAD

**Issue by Wercho** [KSP-KOS/KOS#620]
I get an error anytime I try to run a "LIST ... FROM..." command, for instance:

```
list parts from ship in p1.
Syntax: Unexpected token 'f' found. Expected EOI at line 1
list parts from ship in p1.
           ^_
           VERBOSE DESCRIPTION
The parser used by kOS is complaining about a
part of the script it can't understand.  kOS uses
a parser-generator tool known as TinyPG, and the
text description you see from this error message
comes mostly from TinyPG, not from kOS itself.
Error   Line: 1
Error Column: 12
Message: Unexpected token 'f' found. Expected EOI
list parts rom ship in p1.
```

It appears that kOS no longer understands the "FROM" syntax as documented on the website. I first noticed this in 0.15.6, but I don't know how long ago this happened. It is still the case in 0.16.0. I get the same error if I use a target instead of ship, list engines, dockingports, etc. 

Obviously there are better ways to list the parts from the CPU ship, but I felt that this example was simpler than the real uses for the command, which is listing parts from other (loaded) vessels.

**Comment by erendrake** [comment]
Thank you for the report! I swear I just used a script with it but I will
try to reproduce it in the morning.

**Comment by Dunbaratu** [comment]
I'm not a fan of this syntax, but if the documentation says it works then it needs to be made to work.

In the meantime,

LIST PARTS FROM someVess IN myList.

can be accomplished instead with:

SET myList TO someVess:PARTS.

**Comment by Wercho** [comment]
What is hard to replicate are things like:
LIST dockingports FROM someVess IN myList.

To do that you have to search for them by name, tag them all, or maybe search using someVess:modulesnamed("ModuleDockingNode") and then a FOR loop to get the parts.

Given that no one else has noticed this, I imagine removing the documentation would be a good solution if you'd rather wait to come up with a better syntax. I can work around it.

**Comment by erendrake** [comment]
I am a fan of the syntax, it is easily readable, reasonably terse and extensible. Im sorry that it broke :(

**Comment by Dunbaratu** [comment]
The reason I'm not a fan of it is more how it's implemented.  It puts the responsibility for listing parts of a vessel in the wrong place, outside the vessel class.  If you like the syntax, then LIST FOO FROM BAR IN BAZ should merely be an alias for SET BAZ TO BAR:FOO, generically for any FOO, BAR, and BAZ.  What I think is wrong about it is that there are cases where it's ONLY available using the older syntax and not the newer method calls.

For example LIST DOCKINGPORTS FROM VES IN FOO should be changed to ALSO work as SET FOO TO VES:DOCKINGPORTS.  The LIST BLARG FROM BLAH syntax shouldn't be the only exclusive way to get the list of blargs that are in blah.  That's blah's responsibility to make that available.

The reason I dislike the sytnax is that it's being effectively used as a crutch making us not bother doing it the right way in the class itself.

And remember, if we are serious about things like that .net converter, or that Javascript guy, we can't be taking something that is essentially the class's resposibility to do and making it only available through a kerboscript syntax feature.  If we put it in the class itself, as a suffix call, then it's opened up for other language bindings to come along

**Comment by erendrake** [comment]
I dont think there is any reason not to add dockingport to the vessel but i dont think that either is the right or blessed way, they are just two ways to access the same information. I like

    LIST ENGINES IN elist.   
    vs   
    SET elist to SHIP:ENGINES.
    
    or

    LIST ENGINES FROM 'the other guy' IN elist. 
    vs  
    SET other to VESSEL('the other guy').
    SET elist to other:ENGINES.

It doesnt use the vessel structures directly but who cares. Its just a piece of syntactic sugar. That doesnt make it a crutch or something to be avoided. Its just another tool.

**Comment by Dunbaratu** [comment]
> Its just a piece of syntactic sugar. 

I was advocating for making it BECOME that.  The point is that it would be bad if we implemented dockngports as LIST DOCKINGPORTS FROM VES IN MYLIST. *instead of* as a member method of vessel.  First implement it as a member of vessel, and then make the LIST DOCKINGPORTS be an alias wrapped around that.

"What are the ports in me" is the responsibility of the vessel object, not a third-party outside utility function that has to peek inside a structure it doesn't "own" to calculate the answer.
.

**Comment by erendrake** [comment]
ok, so none of this discussion was germane to fixing the LIST FROM IN syntax? it was a complaint about the vessel structure? alright, then im going to fix this and we can move that discussion to a different issue where it belongs.

**Comment by Dunbaratu** [comment]
My comments were germane to this, and spawned by this:

> What is hard to replicate are things like:
> LIST dockingports FROM someVess IN myList.

The implication of which is "fix it in list because it's not in vessel", which isn't the right way to look at that problem.

**Comment by erendrake** [comment]
i disagree that is it more important. they are the same in my mind because they would be the same tool in two places. Discussion about changing the vessel structure does not belong in here. lets continue it over here: https://github.com/KSP-KOS/KOS/issues/634 rather than arguing about something we agree on in an unrelated issue.

**Comment by abenkovskii** [comment]
If I understand the `list from` documentation right than this:
```
list files from vessel_with_two_kOS_precessors in a.
list targets from my_favorite_asteroid in c.
list bodies from my_minmus_base in d.
```
is a "valid" syntax. I don't like this idea.

**Comment by abenkovskii** [comment]
Let me clarify:
1. I see the problem that the syntax `list foo in bar.` solves
2. I like this piece of syntactic sugar
3. But I don't like `list foo from bar in baz.` because some "list keywords" are not per-vessel

**Comment by abenkovskii** [comment]
Proposal:
Because "list keywords" are already reserved we can:
1. Turn them into bindings returning read only lists
2. Make `list foo in bar.` to be an alias to `local bar is foo:copy().`
3. Make `list foo from vessel in bar.` to be an alias to `local bar is vessel:foo:copy().`
This way we'll avoid having one "god" place that contains all listing logic.

**Comment by abenkovskii** [comment]
The problem of my design: how to threat `list foo.` command?

**Comment by abenkovskii** [comment]
The benefit of proposed design:
```
for file in files {do stuff.}
```
instead of:
```
list files in f_list.
for file in f_list {do stuff.}
```

**Comment by fachrynewtron** [comment]
I have the same issue, I just install KOS 0.172 in KSP 1.02 and get the same TinyPG issue when try to edit

**Comment by abenkovskii** [comment]
> I have the same issue

@fachrynewtron, what exactly you are trying to edit?

**Comment by gisikw** [comment]
Reproduced the same issue - any update on this? (in particular, looking to check that a file exists on a specific volume)
